pep-9999.rst

```diff
@@ -0,0 +1,393 @@
+PEP: 9999
+Title: Type defaults for TypeVars
+Version: $Revision$
+Last-Modified: $Date$
+Author: James Hilton-Balfe <gobot1234yt@gmail.com>
+Sponsor: Jelle Zijlstra <jelle.zijlstra@gmail.com>
+Discussions-To: typing-sig@python.org
+Status: Draft
+Type: Standards Track
+Topic: Typing
+Content-Type: text/x-rst
+Created: 14-Jul-2022
+Python-Version: 3.12
+
+Type defaults for TypeVars
+==========================
+
+Abstract
+--------
+
+This PEP introduces the concept of type defaults for TypeVars, which act
+as defaults for a type parameter when none is specified.
+
+Motivation
+----------
+
+.. code:: py
+
+   T = TypeVar("T", default=int)  # This means that if no type is specified T = int
+
+   @dataclass
+   class Box(Generic[T]):
+       value: T | None = None
+
+   reveal_type(Box())                      # type is Box[int]
+   reveal_type(Box(value="Hello World!"))  # type is Box[str]
+
+One place this `regularly comes
+up <https://github.com/python/typing/issues/975>`__ is ``Generator``. I
+propose changing the *stub definition* to something like:
+
+.. code:: py
+
+   YieldT = TypeVar("YieldT")
+   SendT = TypeVar("SendT", default=None)
+   ReturnT = TypeVar("ReturnT", default=None)
+
+   class Generator(Generic[YieldT, SendT, ReturnT]): ...
+
+   Generator[int] == Generator[int, None] == Generator[int, None, None]
+
+This is also useful for a Generic that is commonly over one type.
+
+.. code:: py
+
+   class Bot: ...
+
+   BotT = TypeVar("BotT", bound=Bot, default=Bot)
+
+   class Context(Generic[BotT]):
+       bot: BotT
+
+   class MyBot(Bot): ...
+
+   reveal_type(Context().bot)         # type is Bot  # notice this is not Any which is what it would be currently
+   reveal_type(Context[MyBot]().bot)  # type is MyBot
+
+Not only does this improve typing for those who explicitly use it. It
+also helps non-typing users who rely on auto-complete to speed up their
+development.
+
+This design pattern is common in projects like:
+ - `discord.py <https://github.com/Rapptz/discord.py>`__ - where the
+   example above was taken from.
+ - `NumPy <https://github.com/numpy/numpy>`__ - the default for types
+   like ``ndarray``\ 's ``dtype`` would be ``float64``. Currently it's
+   ``Unknown`` or ``Any``.
+ - `TensorFlow <https://github.com/tensorflow/tensorflow>`__ (this
+   could be used for Tensor similarly to ``numpy.ndarray`` and would be
+   useful to simplify the definition of ``Layer``)
+
+`This proposal could also be used on builtins.slice <https://github.com/python/typing/issues/159>`__
+where the parameter should ``start`` default to ``int``, ``stop``
+default to ``start`` and step default to ``int | None``
+
+.. code:: py
+
+   StartT = TypeVar("StartT", default=int)
+   StopT = TypeVar("StopT", default=StartT)
```

Comment:
I'm not sure we should allow a TypeVar as the default for another TypeVar; this may make the TypeVar difficult to solve for in some cases. @erictraut do you have any opinion here?

Reply:
This is something that typescript supports so I naively thought it can't be an impossible feature to support

Reply:
I don't think this makes sense. A TypeVar used as a type argument must be associated with a scope (a generic class, function, or type alias). What scope is this type variable associated with? Or are you saying that this would be possible only in cases where the TypeVar is already associated with an outer scope?

```python
class Outer[T]:
    class Inner[X = T]:
        ...
```

I think the use cases for this are very rare, and it adds significant complexity — in the type checker implementation and in the specification (more edge cases to reason about), so I'd recommend against including this.

Reply:
Yes it would need to have the TypeVar it's defaulting to bound to an outer scope so you could also do
```py
class slice[StartT = int, StopT = StartT, StepT = int | None]: ...

slice[str]()  # type is slice[str, str, int | None]
```
(in TS)
```ts
class Slice<StartT = number, StopT = StartT, StepT = number | null> {};

new Slice<string>()  // type is Slice<string, string, number | null>
```
I don't think there are that many edge cases to reason with here, if a TypeVar with a default (T2) that is another TypeVar (T1) then T1 needs to be in scope before T2 can be used otherwise it's an error.

i.e.
```py
class slice[StartT = StopT, StopT = int, StepT = int | None]: ...
```
doesn't work, I think this fits in quite neatly with PEP 695's handling of this.

Reply:
No, this won't work for PEP 695. Because of all the runtime constraints that are imposed on the runtime, I need to allocate all of the type parameters before binding them to their names. That means the TypeVars `StartT`, `StopT`, and `StepT` in your example above need to be constructed prior to referring to any of them by name. So the default expressions won't be able to refer to other type parameters.

I'd really prefer not to remove this from the PEP.

Reply:
I guess we could support this in the PEP 695 implementation if we construct the type variables first, then evaluate the default expressions and "inject" the result into the already-constructed type variable. The equivalent of this...

```python
__local_typevars__ = (TypeVar("StartT"), TypeVar("StopT"), TypeVar("StepT))
__local_typevars__[0].__default__ = int
__local_typevars__[1].__default__ = __local_typevars__[0]
__local_typevars__[2].__default = int | None
```